Core/Src/ws2812.c

```diff
@@ -0,0 +1,213 @@
+			out_buf[i] = 0;
+		} else {
+			out_buf[i] = PWM_LO;
+		}
+
+	}
+}
+
+void led_set_RGB_index(uint8_t index, uint32_t color_code) {
+	//Get the brightness of each color
+	uint8_t green_brightness = get_green_from_hex(color_code);
+	uint8_t red_brightness = get_red_from_hex(color_code);
+	uint8_t blue_brightness = get_blue_from_hex(color_code);
+
+	uint8_t mask = 1;
+
+	for (int g = 7; g >= 0; g--) {
+		if (green_brightness & mask) {
+			out_buf[g + (index+1)*24] = PWM_HI;
+		} else {
+			out_buf[g + (index+1)*24] = PWM_LO;
+		}
+		mask = mask << 1;
+	}
+
+	mask = 1;
+
+	for (int r = 7; r >= 0; r--) {
+		if (red_brightness & mask) {
+			out_buf[r + (index +1)*24 + 8] = PWM_HI;
+		} else {
+			out_buf[r + (index+1)*24 + 8] = PWM_LO;
+		}
+		mask = mask << 1;
+	}
+
+	mask = 1;
+
+	for (int b = 7; b >= 0; b--) {
+		if (blue_brightness & mask) {
+			out_buf[b + (index +1)*24 + 16] = PWM_HI;
+		} else {
+			out_buf[b + (index +1)*24 + 16] = PWM_LO;
+		}
+		mask = mask << 1;
+	}
+
+}
+
+//set the color for all the LEDS
+void led_set_all_RGBs(uint32_t color_code) {
+	for (int i = 0; i < NUM_LEDS; i++) {
+		led_set_RGB_index(i, color_code);
+	}
+}
+
+void led_render_RGB() {
+	//As long as you're not cycling through the LEDs.
+	HAL_TIMEx_PWMN_Start_DMA(&htim1, TIM_CHANNEL_2, (uint32_t*) out_buf,
+					BUFF_SIZE);
+}
+
+//void led_render_cycle(uint32_t color_code) {
+//	for (int i = 0; i < NUM_LEDS; i++) {
+//		led_set_RGB_index(i, color_code);
+//
+//	}
+//
+//}
+//
+
+//void HAL_TIM_PWM_PulseFinishedHalfCpltCallback(TIM_HandleTypeDef *htim) {
+//  // DMA buffer set from LED(wr_buf_p) to LED(wr_buf_p + 1)
+//  if(wr_buf_p < NUM_PIXELS) {
+//    // We're in. Fill the even buffer
+//#if (NUM_BPP == 4) // SK6812
+//    for(uint_fast8_t i = 0; i < 8; ++i) {
+//      wr_buf[i     ] = PWM_LO << (((rgb_arr[4 * wr_buf_p    ] << i) & 0x80) > 0);
+//      wr_buf[i +  8] = PWM_LO << (((rgb_arr[4 * wr_buf_p + 1] << i) & 0x80) > 0);
+//      wr_buf[i + 16] = PWM_LO << (((rgb_arr[4 * wr_buf_p + 2] << i) & 0x80) > 0);
+//      wr_buf[i + 24] = PWM_LO << (((rgb_arr[4 * wr_buf_p + 3] << i) & 0x80) > 0);
+//    }
+//#else // WS2812B
+//    for(uint_fast8_t i = 0; i < 8; ++i) {
+//      wr_buf[i     ] = PWM_LO << (((rgb_arr[3 * wr_buf_p    ] << i) & 0x80) > 0);
+//      wr_buf[i +  8] = PWM_LO << (((rgb_arr[3 * wr_buf_p + 1] << i) & 0x80) > 0);
+//      wr_buf[i + 16] = PWM_LO << (((rgb_arr[3 * wr_buf_p + 2] << i) & 0x80) > 0);
+//    }
+//#endif // End SK6812 WS2812B case differentiation
+//    wr_buf_p++;
+//  } else if (wr_buf_p < NUM_PIXELS + 2) {
+//    // Last two transfers are resets. SK6812: 64 * 1.25 us = 80 us == good enough reset
+//  	//                               WS2812B: 48 * 1.25 us = 60 us == good enough reset
+//    // First half reset zero fill
+//    for(uint8_t i = 0; i < WR_BUF_LEN / 2; ++i) wr_buf[i] = 0;
+//    wr_buf_p++;
+//  }
+//}
+//
+void HAL_TIM_PWM_PulseFinishedCallback(TIM_HandleTypeDef *htim) {
```

Comment:
this is a very empty function - do we still need it? What happens if we remove it entirely?